mina.netty/src/main/java/org/jboss/netty/channel/socket/nio/NioChildDatagramPipelineSink.java

```diff
@@ -0,0 +1,122 @@
+
+    private final WorkerPool<NioDatagramWorker> workerPool;
+
+    /**
+     * Creates a new {@link NioDatagramPipelineSink} with a the number of {@link NioDatagramWorker}s
+     * specified in workerCount.  The {@link NioDatagramWorker}s take care of reading and writing
+     * for the {@link NioDatagramChannel}.
+     *
+     * @param workerExecutor
+     *        the {@link Executor} that will run the {@link NioDatagramWorker}s
+     *        for this sink
+     * @param workerCount
+     *        the number of {@link NioDatagramWorker}s for this sink
+     */
+    NioChildDatagramPipelineSink(final WorkerPool<NioDatagramWorker> workerPool) {
+        this.workerPool = workerPool;
+    }
+
+    /**
+     * Handle downstream event.
+     *
+     * @param pipeline the {@link ChannelPipeline} that passes down the
+     *                 downstream event.
+     * @param e The downstream event.
+     */
+    public void eventSunk(final ChannelPipeline pipeline, final ChannelEvent e)
+            throws Exception {
+
+        final NioDatagramChannel channel = (NioDatagramChannel) e.getChannel();
+        final ChannelFuture future = e.getFuture();
+        if (e instanceof ChannelStateEvent) {
+            final ChannelStateEvent stateEvent = (ChannelStateEvent) e;
+            final ChannelState state = stateEvent.getState();
+            final Object value = stateEvent.getValue();
+            switch (state) {
+                case OPEN:
+                    if (Boolean.FALSE.equals(value)) {
+                        channel.worker.close(channel, future);
+                    }
+                    break;
+            }
+        } else if (e instanceof MessageEvent) {
+            // final MessageEvent event = (MessageEvent) e;
+            // final boolean offered = parent.writeBufferQueue.offer(event);
+            // assert offered;
+            // parent.worker.writeFromUserCode(parent);
+            //
+            // In netty AbstractNioChannel#writeBufferQueue is thread-safe but not in mina.netty's copy.
+            // Scheduling the write on parent channel thread since child channels and parent channel are
+            // on different i/o threads
+            NioDatagramChannel parent = (NioDatagramChannel) channel.getParent();
+
+            class WriteTask implements Runnable {
+                @Override
+                public void run() {
+                    Object message = ((MessageEvent) e).getMessage();
+                    ChannelFuture parentFuture = parent.write(message, channel.getRemoteAddress());
+                    parentFuture.addListener(f -> {
+                        if (f.isSuccess()) {
+                            future.setSuccess();
+                        } else {
+                            future.setFailure(f.getCause());
+                        }
+                    });
+                }
+            }
+
+            parent.getWorker().executeInIoThread(new WriteTask());
```

Comment:
Here there is object allocation for each write. I will create a lambda for writeTask and that doesn't an object.